### PR TITLE
Implement Multimap.collectKeyMultiValues(). Fixes #546

### LINKED ISSUE #546: Implement Multimap.collectKeyMultiValues().
I've written this as static utility twice. I feel like it belongs in the framework. Thoughts?

```java
public static <InputKey, InputValue, OutputKey, OutputValue> ImmutableListMultimap<OutputKey, OutputValue> collectKeyMultiValues(
        MutableListMultimap<InputKey, InputValue> multimap,
        Function<? super InputKey, ? extends OutputKey> keyFunction,
        Function<? super InputValue, ? extends OutputValue> valueFunction)
{
    MutableListMultimap<OutputKey, OutputValue> result = Multimaps.mutable.list.empty();
    multimap.forEachKeyMultiValues((key, multiValues) ->
            result.putAll(
                    keyFunction.valueOf(key),
                    Iterate.collect(multiValues, valueFunction)));
    return result.toImmutable();
}
```

### PR DESCRIPTION
**Pull request by nikhilnanivadekar**

### CODE REVIEW

File: unit-tests/src/test/java/org/eclipse/collections/impl/multimap/AbstractImmutableMultimapTestCase.java [hunk 1]
```diff
@@ -205,6 +205,14 @@ public void collectKeysValues()
         Assert.assertEquals(this.classUnderTest().newWith("OneKey", "1Value").newWith("TwoKey", "2Value"), collectedMultimap);
     }
 
+    @Test
+    public void collectKeyMultiValues()
+    {
+        Multimap<String, String> multimap = this.<String, String>classUnderTest().newWith("One", "1").newWith("Two", "2");
+        Multimap<String, String> collectedMultimap = multimap.collectKeyMultiValues(argument1 -> argument1 + "Key", argument2 -> argument2 + "Value");
+        Assert.assertEquals(this.classUnderTest().newWith("OneKey", "1Value").newWith("TwoKey", "2Value"), collectedMultimap);
+    }
```
[the remark this answers is not in the document]
Comment (nikhilnanivadekar) [reply unit-tests/src/test/java/org/eclipse/collections/impl/multimap/AbstractImmutableMultimapTestCase.java]: these assertions are in the sub-classes no?

**Comment by nikhilnanivadekar** [comment]
I have added the `since` annotation as well as fixed the efficiency concern. Happy to add more tests, but I went through the existing tests and it seems that I have covered most cases.

**Comment by nikhilnanivadekar** [comment]
I’ll fix that. Any other comments or it’s good to go after that?

**Comment by motlin** [comment]
I think that's it. It was already pretty much good to go.

**Comment by nikhilnanivadekar** [comment]
All builds have passed. It is rebased on to master. Do the honors!

**Comment by nikhilnanivadekar** [comment]
@motlin please review and merge if everything looks ok

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/multimap/Multimap.java [hunk 2]
```diff
@@ -455,6 +455,36 @@
...
+     * @param valueFunction {@link Function} to use transformation to get the values
+     * @return a new {@code Multimap}, which contains elements as a result of the transformation
+     * @since 10.0
+     */
+    <K2, V2> Multimap<K2, V2> collectKeyMultiValues(Function<? super K, ? extends K2> keyFunction, Function<? super V, ? extends V2> valueFunction);
+
+    /**
+     * Same as the collectKeyMultiValues method but uses the specified target multimap for the results.
+     * <p>
+     * <pre>e.g.
```
Comment (motlin) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/multimap/Multimap.java]: Both "e.g."s should be outside the pre-formatted sections

Comment (nikhilnanivadekar) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/multimap/Multimap.java]: It’s consistent with rest of the file and the build is going through fine also when I ctrl+q in IntelliJ the javadoc looks fine

Comment (motlin) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/multimap/Multimap.java]: I'm not saying the Javadoc will break, I'm saying it's conceptually wrong. It's formatted as code, but it's part of the prose.

Comment (nikhilnanivadekar) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/multimap/Multimap.java]: I can have another commit fixing all this. But I don’t think that is critical right now.

Comment (nikhilnanivadekar) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/multimap/Multimap.java]: I can have another commit fixing all this. But I don’t think that is critical right now.

**Pull request merged**